File: packages/butler-agent/src/agent/cognition/memory/recall/interpretations.ts
import type { Database } from "bun:sqlite";
import type { RecallMemoryInput, RecallMemoryResult } from "./contracts.ts";
import { readClaim } from "../projection/claim-store.ts";
import { rawMemorySourceId } from "./source-ref.ts";

type Interpretation = NonNullable<RecallMemoryResult["results"][number]["interpretations"]>[number];
/** Attach interpretations only to evidence actually delivered by the public recall path. */
export function resultInterpretations(db: Database, input: RecallMemoryInput, evidence: RecallMemoryResult["results"][number]["evidence"]): Interpretation[] {
  const refs = new Map(evidence.map((item) => [rawMemorySourceId(item.source_ref), item.source_ref]));
  if (!refs.size) return [];
  const nodes = db.query<{ node_id: string }, any>(`SELECT DISTINCT c.node_id FROM memory_claims c
    JOIN memory_evidence e ON e.node_id=c.node_id WHERE e.source_id IN (${[...refs].map(() => "?").join(",")}) ORDER BY c.node_id LIMIT 8`).all(...refs.keys());
  return nodes.flatMap(({ node_id }) => {
    const claim = readClaim(db, node_id)!;
    const dependencies = db.query<{ source_id: string }, [string]>("SELECT source_id FROM memory_evidence WHERE node_id=?").all(node_id);
    const sources = dependencies.flatMap((row) => refs.has(row.source_id) ? [refs.get(row.source_id)!] : []);
    if (sources.length !== dependencies.length) return [];
    let status: Interpretation["status"] = (claim.valid_from && Date.parse(claim.valid_from) > Date.parse(input.asOf)) ||
      (claim.valid_to && Date.parse(claim.valid_to) <= Date.parse(input.asOf)) ? "historical" : "recorded";
    const changes = db.query<{ rel_type: string }, any>(`SELECT DISTINCT e.rel_type FROM edges e
      JOIN edge_evidence ee ON ee.edge_id=e.edge_id JOIN memory_chunk_sources s ON s.source_id=ee.chunk_source_id
      JOIN memory_chunks c ON c.memory_chunk_id=s.episode_id AND c.current_revision=s.revision
      WHERE e.target_node_id=? AND e.status='active' AND c.status='active'
        AND ee.chunk_source_id IN (${[...refs].map(() => "?").join(",")})
        AND e.rel_type IN ('supersedes','contradicts','refines')`).all(node_id, ...refs.keys());
    if (changes.some((row) => row.rel_type === "refines")) status = "refined";
    if (changes.some((row) => row.rel_type === "contradicts")) status = "conflicted";
    if (changes.some((row) => row.rel_type === "supersedes")) status = "superseded";
    return [{ node_ref: node_id, statement: claim.statement, speech_act: claim.speech_act, basis: claim.basis,
      source_class: claim.source_class, authority: claim.authority, source_refs: sources,
      support_complete: sources.length === dependencies.length, status }];
  });
}

File: packages/butler-agent/rust/agent/src/cognition/graph/recall/claims/interpretations.rs
use std::collections::{HashMap, HashSet};

use rusqlite::{Connection, params_from_iter, types::Value};

use crate::cognition::{
    CognitionResult,
    recall::{InterpretationStatus, RecallInterpretation, RecallRequest},
};

use super::super::{db_error, scope};

pub(super) fn interpretations(
    db: &Connection,
    input: &RecallRequest,
    refs: &[(String, String)],
    parse_date: impl Fn(&str) -> f64,
) -> CognitionResult<Vec<RecallInterpretation>> {
    if refs.is_empty() {
        return Ok(Vec::new());
    }
    let refs = refs.iter().cloned().collect::<HashMap<_, _>>();
    let ids = refs.keys().cloned().collect::<Vec<_>>();
    let sql = format!(
        r"
        SELECT DISTINCT c.node_id FROM memory_claims c
        JOIN memory_evidence e ON e.node_id=c.node_id
        WHERE e.source_id IN ({}) ORDER BY c.node_id LIMIT 8
    ",
        scope::placeholders(ids.len())
    );
    let mut statement = db.prepare(&sql).map_err(db_error)?;
    let nodes = statement
        .query_map(
            params_from_iter(ids.iter().cloned().map(Value::Text)),
            |row| row.get::<_, String>(0),
        )
        .map_err(db_error)?
        .collect::<Result<Vec<_>, _>>()
        .map_err(db_error)?;
    let mut claims = db
        .prepare(
            "SELECT statement,speech_act,basis,source_class,authority,valid_from,valid_to
         FROM memory_claims WHERE node_id=?",
        )
        .map_err(db_error)?;
    let mut dependencies = db
        .prepare("SELECT source_id FROM memory_evidence WHERE node_id=?")
        .map_err(db_error)?;
    let changes_sql = format!(
        r"
        SELECT DISTINCT e.rel_type FROM edges e
        JOIN edge_evidence ee ON ee.edge_id=e.edge_id
        JOIN memory_chunk_sources s ON s.source_id=ee.chunk_source_id
        JOIN memory_chunks c ON c.memory_chunk_id=s.episode_id
          AND c.current_revision=s.revision
        WHERE e.target_node_id=? AND e.status='active' AND c.status='active'
          AND ee.chunk_source_id IN ({})
          AND e.rel_type IN ('supersedes','contradicts','refines')
    ",
        scope::placeholders(ids.len())
    );
    let mut changes = db.prepare(&changes_sql).map_err(db_error)?;
    let mut output = Vec::new();
    for node_id in nodes {
        let (statement, speech_act, basis, source_class, authority, valid_from, valid_to): (
            String,
            String,
            String,
            String,
            String,
            Option<String>,
            Option<String>,
        ) = claims
            .query_row([&node_id], |row| {
                Ok((
                    row.get(0)?,
                    row.get(1)?,
                    row.get(2)?,
                    row.get(3)?,
                    row.get(4)?,
                    row.get(5)?,
                    row.get(6)?,
                ))
            })
            .map_err(db_error)?;
        let dependent = dependencies
            .query_map([&node_id], |row| row.get::<_, String>(0))
            .map_err(db_error)?
            .collect::<Result<Vec<_>, _>>()
            .map_err(db_error)?;
        let source_refs = dependent
            .iter()
            .filter_map(|id| refs.get(id).cloned())
            .collect::<Vec<_>>();
        if source_refs.len() != dependent.len() {
            continue;
        }
        let mut status = if valid_from
            .as_deref()
            .is_some_and(|from| parse_date(from) > parse_date(&input.as_of))
            || valid_to
                .as_deref()
                .is_some_and(|to| parse_date(to) <= parse_date(&input.as_of))
        {
            InterpretationStatus::Historical
        } else {
            InterpretationStatus::Recorded
        };
        let mut args = vec![Value::Text(node_id.clone())];
        args.extend(ids.iter().cloned().map(Value::Text));
        let relations = changes
            .query_map(params_from_iter(args), |row| row.get::<_, String>(0))
            .map_err(db_error)?
            .collect::<Result<HashSet<_>, _>>()
            .map_err(db_error)?;
        if relations.contains("refines") {
            status = InterpretationStatus::Refined;
        }
        if relations.contains("contradicts") {
            status = InterpretationStatus::Conflicted;
        }
        if relations.contains("supersedes") {
            status = InterpretationStatus::Superseded;
        }
        output.push(RecallInterpretation {
            node_ref: node_id,
            statement,
            speech_act,
            basis,
            source_class,
            authority,
            source_refs,
            support_complete: true,
            status,
        });
    }
    Ok(output)
}
